File: 01-Languages/Go/17-API-Integration/main.go
// main.go - the built-in net/http client (no dependency needed, unlike Java/C++), plus
// encoding/json (Lesson 10) for decoding the response.
package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
)

type Todo struct {
	UserID    int    `json:"userId"`
	ID        int    `json:"id"`
	Title     string `json:"title"`
	Completed bool   `json:"completed"`
}

func main() {
	fmt.Println("--- GET https://jsonplaceholder.typicode.com/todos/1 ---")
	resp, err := http.Get("https://jsonplaceholder.typicode.com/todos/1")
	if err != nil {
		panic(err)
	}
	defer resp.Body.Close()

	fmt.Println("status:", resp.StatusCode)
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		panic(err)
	}
	var todo Todo
	if err := json.Unmarshal(body, &todo); err != nil {
		panic(err)
	}
	fmt.Printf("Decoded: UserID=%d, Title=%s, Completed=%t\n", todo.UserID, todo.Title, todo.Completed)

	fmt.Println("\n--- GET a route that returns 404 ---")
	resp2, err := http.Get("https://jsonplaceholder.typicode.com/todos/99999999")
	if err != nil {
		panic(err)
	}
	defer resp2.Body.Close()
	fmt.Println("status:", resp2.StatusCode)
	fmt.Println("net/http does NOT return an error for a 404 -- StatusCode must be checked manually.")

	fmt.Println("\n--- POST with a JSON body ---")
	newTodo := map[string]interface{}{
		"title":     "Learn net/http",
		"completed": false,
		"userId":    1,
	}
	jsonBody, err := json.Marshal(newTodo)
	if err != nil {
		panic(err)
	}
	resp3, err := http.Post(
		"https://jsonplaceholder.typicode.com/todos",
		"application/json",
		bytes.NewBuffer(jsonBody),
	)
	if err != nil {
		panic(err)
	}
	defer resp3.Body.Close()
	fmt.Println("status:", resp3.StatusCode)
	respBody, _ := io.ReadAll(resp3.Body)
	fmt.Println("body (echoed back with a fake id):", string(respBody))
}

File: 01-Languages/Rust/17-API-Integration/src/main.rs
// main.rs - HTTP requests with the `ureq` crate (since std provides no HTTP client at all),
// plus serde/serde_json (Lesson 10) for decoding the response into a struct.

use serde::{Deserialize, Serialize};

#[derive(Debug, Deserialize)]
struct Todo {
    #[serde(rename = "userId")]
    user_id: i32,
    id: i32,
    title: String,
    completed: bool,
}

#[derive(Serialize)]
struct NewTodo {
    title: String,
    completed: bool,
    #[serde(rename = "userId")]
    user_id: i32,
}

fn main() {
    println!("--- GET https://jsonplaceholder.typicode.com/todos/1 ---");
    let response = ureq::get("https://jsonplaceholder.typicode.com/todos/1").call();
    match response {
        Ok(resp) => {
            println!("status: {}", resp.status());
            let todo: Todo = resp.into_json().expect("failed to parse JSON");
            println!(
                "Decoded: id={}, user_id={}, title={}, completed={}",
                todo.id, todo.user_id, todo.title, todo.completed
            );
        }
        Err(e) => println!("request error: {}", e),
    }

    println!("\n--- GET a route that returns 404 ---");
    // ureq treats non-2xx as an Err(ureq::Error::Status(...)) by default -- unlike every
    // other language course's HTTP client, which returns a normal, non-error response object.
    match ureq::get("https://jsonplaceholder.typicode.com/todos/99999999").call() {
        Ok(resp) => println!("status: {}", resp.status()),
        Err(ureq::Error::Status(code, _)) => {
            println!("status: {} (returned as an Err by ureq's design, unlike most other clients)", code);
        }
        Err(e) => println!("other error: {}", e),
    }

    println!("\n--- POST with a JSON body ---");
    let new_todo = NewTodo {
        title: "Learn ureq".to_string(),
        completed: false,
        user_id: 1,
    };
    match ureq::post("https://jsonplaceholder.typicode.com/todos").send_json(&new_todo) {
        Ok(resp) => {
            println!("status: {}", resp.status());
            let body: serde_json::Value = resp.into_json().expect("failed to parse JSON");
            println!("body (echoed back with a fake id): {}", body);
        }
        Err(e) => println!("request error: {}", e),
    }
}
